File: SupportPreference2Demo/app/src/main/java/edu/cs4730/supportpreference2demo/MainActivity.java
package edu.cs4730.supportpreference2demo;

import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.graphics.Insets;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;
import androidx.fragment.app.Fragment;
import androidx.preference.Preference;
import androidx.preference.PreferenceFragmentCompat;

import edu.cs4730.supportpreference2demo.databinding.ActivityMainBinding;

/**
 * This is a simple example of using the support library PreferenceFragmentCompat.
 * The MainFragment has a button to launch the SettingsFragment which shows the
 * preferences, which are saved automatically to SharedPreferences.
 * <p>
 * It can't use the Arch Navigation component because it doesn't support
 * PreferenceFragmentCompat (as of this writing anyway 9/10/2025 ).  So the code is little gross in main.
 * <p>
 * It appears M3 has a preference theme built in, so no custom theme is needed like in SupportPreferenceDemo.
 */

public class MainActivity extends AppCompatActivity implements MainFragment.OnFragmentInteractionListener,
    PreferenceFragmentCompat.OnPreferenceStartFragmentCallback {

    ActivityMainBinding binding;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityMainBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());
        ViewCompat.setOnApplyWindowInsetsListener(binding.container, (v, insets) -> {
            Insets systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars());
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom);
            return WindowInsetsCompat.CONSUMED;
        });
        getSupportFragmentManager().beginTransaction().add(binding.container.getId(), new MainFragment()).commit();
    }

    /**
     * Called when the user has clicked on a preference that has a fragment class name
     * associated with it. The implementation should instantiate and switch to an instance
     * of the given fragment.
     *
     * @param caller The fragment requesting navigation
     * @param pref   The preference requesting the fragment
     * @return {@code true} if the fragment creation has been handled
     */
    @Override
    public boolean onPreferenceStartFragment(@NonNull PreferenceFragmentCompat caller, Preference pref) {

        //take from https://developer.android.com/develop/ui/views/components/settings/organize-your-settings

        // Instantiate the new Fragment.
        final Bundle args = pref.getExtras();
        final Fragment fragment = getSupportFragmentManager().getFragmentFactory().instantiate(
            getClassLoader(), pref.getFragment());
        fragment.setArguments(args);

        // Replace the existing Fragment with the new Fragment.
        getSupportFragmentManager().beginTransaction().replace(binding.container.getId(), fragment)
            .addToBackStack(null).commit();
        return true;
    }

    @Override
    public void onFragmentInteraction(int which) {
        getSupportFragmentManager().beginTransaction().replace(binding.container.getId(), new SettingsFragment()).addToBackStack(null).commit();
    }
}

File: SupportPreference2Demo_kt/app/src/main/java/edu/cs4730/supportpreference2demo_kt/MainActivity.kt
package edu.cs4730.supportpreference2demo_kt

import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.preference.Preference
import androidx.preference.PreferenceFragmentCompat
import edu.cs4730.supportpreference2demo_kt.databinding.ActivityMainBinding

/**
 * This is a simple demo of the AndroidX Preference library, which is the replacement for
 * the deprecated Android Preference library.  This demo is similar to the original demo,
 * but uses only fragments, not activities.  So everything is done in MainActivity and
 * fragments.
 *
 * Note: The design support library version of preference is no longer being maintained.
 * Google suggests using the AndroidX Preference library instead, which is what this demo
 * shows.
 * 
 * lastly, Arch Navigation component can be used with preference as well, since it doesn't support
 * the PreferenceFragmentCompat.OnPreferenceStartFragmentCallback interface. So the code is little gross.
 *
 * Note: this example is the same as the java version, just in kotlin.  Except for one weird thing
 * in SettingsFragment, I give up on DropDownPreference SimpleSummaryProvider is broken in kotlin for some reason.
 */

class MainActivity : AppCompatActivity(), MainFragment.OnFragmentInteractionListener,
    PreferenceFragmentCompat.OnPreferenceStartFragmentCallback {

    lateinit var binding: ActivityMainBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.container)) { v, insets ->
            val systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars())
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom)
            WindowInsetsCompat.CONSUMED
        }
        supportFragmentManager.beginTransaction()
            .add(binding.container.getId(), MainFragment()).commit()
    }

    /**
     * Called when the user has clicked on a preference that has a fragment class name
     * associated with it. The implementation should instantiate and switch to an instance
     * of the given fragment.
     *
     * @param caller The fragment requesting navigation
     * @param pref   The preference requesting the fragment
     * @return `true` if the fragment creation has been handled
     */
    override fun onPreferenceStartFragment(
        caller: PreferenceFragmentCompat,
        pref: Preference
    ): Boolean {
        //take from https://developer.android.com/develop/ui/views/components/settings/organize-your-settings

        val args = pref.getExtras()
        val fragment = supportFragmentManager.getFragmentFactory().instantiate(
            classLoader, pref.fragment!!
        )
        fragment.setArguments(args)

        // Replace the existing Fragment with the new Fragment.
        supportFragmentManager.beginTransaction().replace(binding.container.getId(), fragment)
            .addToBackStack(null).commit()
        return true
    }

    override fun onFragmentInteraction(which: Int) {
        supportFragmentManager.beginTransaction()
            .replace(binding.container.getId(), SettingsFragment()).addToBackStack(null).commit()
    }

}